lib/rspec-puppet/support.rb

```diff
@@ -217,7 +217,23 @@ def with_vardir
       end
     end
 
+    def reset_providers
+      Puppet::Type.eachtype do |type|
+        # Unset the default provider for each type as this is cached in a class
+        # instance method and so doesn't get re-evaluated when we change facts
+        # and compile a new catalogue.
+        type.defaultprovider = nil
+      end
+    end
+
     def build_catalog_without_cache(nodename, facts_val, trusted_facts_val, hiera_config_val, code, exported)
+      # Blindly stub out any feature and command Puppet::Confine evaluations
+      # that providers might do. This is necessary because we convert the
+      # resource catalogue to a RAL catalogue in order to evaluate all the
+      # automatic graph edges, which requires evaluating the providers.
+      allow_any_instance_of(Puppet::Confine::Feature).to receive(:pass?).with(anything).and_return(true)
+      allow_any_instance_of(Puppet::Util).to receive(:which).with(anything).and_return(true)
+      allow_any_instance_of(Puppet::Provider::Command).to receive(:execute).with(any_args).and_return("")
```

Comment:
The second and third here look like they could massively mess with people's tests.

* Direct `Puppet::Util.which` stubs, probably ok:
  * https://github.com/puppetlabs/puppetlabs-firewall/blob/cbbd9781164351d38cb8391096b8817e5fa8bf9b/spec/unit/puppet/provider/iptables_spec.rb#L32
  * https://github.com/puppetlabs/puppetlabs-mysql/blob/ef7e882ebaf98d5ccd0c16891a396db9f4700bc0/spec/unit/puppet/provider/mysql_user/mysql_spec.rb#L87
  * https://github.com/puppetlabs/puppetlabs-sqlserver/blob/master/spec/unit/puppet/sqlserver_spec_helper.rb#L6

* low-level stubs for `Puppet::Util::Execution.execute`, which will be skipped (through https://github.com/puppetlabs/puppet/blob/801542e53b6f7e7d5edcaea46590139cb42a2949/lib/puppet/provider.rb#L257)
  * https://github.com/puppetlabs/puppetlabs-chocolatey/blob/41a4307ed938a2c5edc1dac78fe8b64a4f80de4e/spec/unit/puppet/provider/chocolateyconfig/windows_spec.rb#L234
  * also in puppetlabs-firewall, puppetlabs-dism, puppetlabs-vcsrepo

As an alternative, maybe we can do it like e.g. stdlib and stub `defaultprovider`: https://github.com/puppetlabs/puppetlabs-stdlib/blob/08d96808c63535da8280477a01631c9b67463595/spec/unit/facter/service_provider_spec.rb#L13 :
```
Puppet::Type.type(:service).stubs(:defaultprovider).returns provider
```

We talked about how that will impact some type's autorequires, and I'm wondering how provider's even would do that during the compile on the server. A quick review of our supported modules showed only acl does that (https://github.com/puppetlabs/puppetlabs-acl/blob/bec0f1634035785b5f9837b64c9adfe62546e7cb/lib/puppet/type/acl.rb#L394) and I'm very sceptical that this works outside a local `puppet apply` on a windows host, as there would be no access to the AD from a linux puppet master.